diff -ur trunk-stock/ffmpeg/libavcodec/arm/mpegvideo_iwmmxt.c trunk/ffmpeg/libavcodec/arm/mpegvideo_iwmmxt.c
--- trunk-stock/ffmpeg/libavcodec/arm/mpegvideo_iwmmxt.c	2011-08-26 13:32:05.163849841 +0800
+++ trunk/ffmpeg/libavcodec/arm/mpegvideo_iwmmxt.c	2011-08-26 13:34:14.603849729 +0800
@@ -112,6 +112,7 @@
 
 void MPV_common_init_iwmmxt(MpegEncContext *s)
 {
+    int mm_flags = av_get_cpu_flags();
     if (!(mm_flags & AV_CPU_FLAG_IWMMXT)) return;
 
     s->dct_unquantize_h263_intra = dct_unquantize_h263_intra_iwmmxt;
