From ee21355fce070d3bb904ba15810b5a4448f01c4b Mon Sep 17 00:00:00 2001
From: Sean Cross <sean@chumby.com>
Date: Sun, 10 Apr 2011 07:10:26 +0800
Subject: [PATCH] pm: Ignore spurrious power switch

---
 arch/arm/mach-mx28/pm.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx28/pm.c b/arch/arm/mach-mx28/pm.c
index 3dcbb1e..c53d2b2 100644
--- a/arch/arm/mach-mx28/pm.c
+++ b/arch/arm/mach-mx28/pm.c
@@ -271,6 +271,10 @@ static 	struct delayed_work pswitch_work;
 static void pswitch_check_work(struct work_struct *work)
 {
 	int pin_value, i;
+
+	pr_info("ignoring pswitch\n");
+	return;
+#if 0
 	for (i = 0; i < PSWITCH_POWER_DOWN_DELAY; i++) {
 		pin_value = __raw_readl(REGS_POWER_BASE + HW_POWER_STS) &
 			BF_POWER_STS_PSWITCH(0x1);
@@ -297,6 +301,7 @@ static void pswitch_check_work(struct work_struct *work)
 		REGS_POWER_BASE + HW_POWER_CTRL_CLR);
 	__raw_writel(BM_POWER_CTRL_PSWITCH_IRQ,
 		REGS_POWER_BASE + HW_POWER_CTRL_CLR);
+#endif
 }
 
 
-- 
1.7.3.5

