diff -Nur linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk.c linux-2.6.35.3/arch/arm/mach-mx28/mx28evk.c
--- linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk.c	2011-06-22 16:55:59.645947494 +0800
+++ linux-2.6.35.3/arch/arm/mach-mx28/mx28evk.c	2011-07-13 13:21:35.082774148 +0800
@@ -150,6 +150,11 @@
 
 	mx28_gpio_init();
 	mx28evk_pins_init();
+
+#if defined(CONFIG_FB_MXS) || defined(CONFIG_FB_MXS_MODULE)
+  mx28evk_lcd_pins_init();
+#endif
+
 	mx28_device_init();
 	mx28evk_device_init();
 }
diff -Nur linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk.h linux-2.6.35.3/arch/arm/mach-mx28/mx28evk.h
--- linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk.h	2011-06-22 16:52:21.860447989 +0800
+++ linux-2.6.35.3/arch/arm/mach-mx28/mx28evk.h	2011-07-13 13:10:37.257375743 +0800
@@ -24,5 +24,8 @@
 void mx28evk_enet_io_lowerpower_enter(void);
 void mx28evk_enet_io_lowerpower_exit(void);
 
+#if defined(CONFIG_FB_MXS) || defined(CONFIG_FB_MXS_MODULE) || defined(CONFIG_FB_CHUMBYFBFW)
+extern void __init mx28evk_lcd_pins_init(void);
+#endif
 
 #endif /* __ASM_ARM_MACH_MX28EVK_H */
diff -Nur linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk_pins.c linux-2.6.35.3/arch/arm/mach-mx28/mx28evk_pins.c
--- linux-2.6.35.3.lastpatch/arch/arm/mach-mx28/mx28evk_pins.c	2011-06-22 16:56:03.027637409 +0800
+++ linux-2.6.35.3/arch/arm/mach-mx28/mx28evk_pins.c	2011-07-13 13:11:08.172824961 +0800
@@ -183,7 +183,211 @@
 	 .drive	= 1,
 	 },
 #endif
+#if defined(CONFIG_MMC_MXS) || defined(CONFIG_MMC_MXS_MODULE)
+	/* Configurations of SSP0 SD/MMC port pins */
+	{
+	 .name	= "SSP0_DATA0",
+	 .id	= PINID_SSP0_DATA0,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA1",
+	 .id	= PINID_SSP0_DATA1,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA2",
+	 .id	= PINID_SSP0_DATA2,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA3",
+	 .id	= PINID_SSP0_DATA3,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA4",
+	 .id	= PINID_SSP0_DATA4,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA5",
+	 .id	= PINID_SSP0_DATA5,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA6",
+	 .id	= PINID_SSP0_DATA6,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DATA7",
+	 .id	= PINID_SSP0_DATA7,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_CMD",
+	 .id	= PINID_SSP0_CMD,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SSP0_DETECT",
+	 .id	= PINID_SSP0_DETECT,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 0,
+	 .drive 	= 1,
+	 .pull 		= 0,
+	 },
+	{
+	 .name	= "SSP0_SCK",
+	 .id	= PINID_SSP0_SCK,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 0,
+	 .drive 	= 2,
+	 .pull 		= 0,
+	 },
+#endif
+#if defined(CONFIG_LEDS_MXS) || defined(CONFIG_LEDS_MXS_MODULE)
+	{
+	 .name = "LEDS_PWM0",
+	 .id = PINID_AUART1_RX,
+	 .fun           = PIN_FUN3,
+	 .strength      = PAD_8MA,
+	 .voltage       = PAD_3_3V,
+	 .pullup        = 1,
+	 .drive         = 1,
+	 .pull          = 1,
+	 },
+	{
+	 .name = "LEDS_PWM1",
+	 .id = PINID_AUART1_TX,
+	 .fun           = PIN_FUN3,
+	 .strength      = PAD_8MA,
+	 .voltage       = PAD_3_3V,
+	 .pullup        = 1,
+	 .drive         = 1,
+	 .pull          = 1,
+	 },
+#endif
+#if defined(CONFIG_SND_MXS_SOC_DAI) || defined(CONFIG_SND_MXS_SOC_DAI_MODULE)
+	/* Configurations of SAIF0 port pins */
+	{
+	 .name	= "SAIF0_MCLK",
+	 .id	= PINID_SAIF0_MCLK,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SAIF0_LRCLK",
+	 .id	= PINID_SAIF0_LRCLK,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SAIF0_BITCLK",
+	 .id	= PINID_SAIF0_BITCLK,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SAIF0_SDATA0",
+	 .id	= PINID_SAIF0_SDATA0,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+	{
+	 .name	= "SAIF1_SDATA0",
+	 .id	= PINID_SAIF1_SDATA0,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	 },
+#endif
+#if defined(CONFIG_SND_SOC_MXS_SPDIF) || \
+       defined(CONFIG_SND_SOC_MXS_SPDIF_MODULE)
+	{
+	 .name	= "SPDIF",
+	 .id	= PINID_SPDIF,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_12MA,
+	 .voltage	= PAD_3_3V,
+	 .pullup	= 1,
+	 .drive 	= 1,
+	 .pull 		= 1,
+	},
+#endif
+};
+
 #if defined(CONFIG_FB_MXS) || defined(CONFIG_FB_MXS_MODULE) || defined(CONFIG_FB_CHUMBYFBFW)
+static struct pin_desc mx28evk_lcd_pins[] = {
 	{
 	 .name  = "LCD_D00",
 	 .id	= PINID_LCD_D00,
@@ -471,209 +675,8 @@
 	 .voltage = PAD_3_3V,
 	 .drive	= 1,
 	 },
-#endif
-#if defined(CONFIG_MMC_MXS) || defined(CONFIG_MMC_MXS_MODULE)
-	/* Configurations of SSP0 SD/MMC port pins */
-	{
-	 .name	= "SSP0_DATA0",
-	 .id	= PINID_SSP0_DATA0,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA1",
-	 .id	= PINID_SSP0_DATA1,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA2",
-	 .id	= PINID_SSP0_DATA2,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA3",
-	 .id	= PINID_SSP0_DATA3,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA4",
-	 .id	= PINID_SSP0_DATA4,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA5",
-	 .id	= PINID_SSP0_DATA5,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA6",
-	 .id	= PINID_SSP0_DATA6,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DATA7",
-	 .id	= PINID_SSP0_DATA7,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_CMD",
-	 .id	= PINID_SSP0_CMD,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SSP0_DETECT",
-	 .id	= PINID_SSP0_DETECT,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_8MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 0,
-	 .drive 	= 1,
-	 .pull 		= 0,
-	 },
-	{
-	 .name	= "SSP0_SCK",
-	 .id	= PINID_SSP0_SCK,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 0,
-	 .drive 	= 2,
-	 .pull 		= 0,
-	 },
-#endif
-#if defined(CONFIG_LEDS_MXS) || defined(CONFIG_LEDS_MXS_MODULE)
-	{
-	 .name = "LEDS_PWM0",
-	 .id = PINID_AUART1_RX,
-	 .fun           = PIN_FUN3,
-	 .strength      = PAD_8MA,
-	 .voltage       = PAD_3_3V,
-	 .pullup        = 1,
-	 .drive         = 1,
-	 .pull          = 1,
-	 },
-	{
-	 .name = "LEDS_PWM1",
-	 .id = PINID_AUART1_TX,
-	 .fun           = PIN_FUN3,
-	 .strength      = PAD_8MA,
-	 .voltage       = PAD_3_3V,
-	 .pullup        = 1,
-	 .drive         = 1,
-	 .pull          = 1,
-	 },
-#endif
-#if defined(CONFIG_SND_MXS_SOC_DAI) || defined(CONFIG_SND_MXS_SOC_DAI_MODULE)
-	/* Configurations of SAIF0 port pins */
-	{
-	 .name	= "SAIF0_MCLK",
-	 .id	= PINID_SAIF0_MCLK,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SAIF0_LRCLK",
-	 .id	= PINID_SAIF0_LRCLK,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SAIF0_BITCLK",
-	 .id	= PINID_SAIF0_BITCLK,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SAIF0_SDATA0",
-	 .id	= PINID_SAIF0_SDATA0,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-	{
-	 .name	= "SAIF1_SDATA0",
-	 .id	= PINID_SAIF1_SDATA0,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	 },
-#endif
-#if defined(CONFIG_SND_SOC_MXS_SPDIF) || \
-       defined(CONFIG_SND_SOC_MXS_SPDIF_MODULE)
-	{
-	 .name	= "SPDIF",
-	 .id	= PINID_SPDIF,
-	 .fun	= PIN_FUN1,
-	 .strength	= PAD_12MA,
-	 .voltage	= PAD_3_3V,
-	 .pullup	= 1,
-	 .drive 	= 1,
-	 .pull 		= 1,
-	},
-#endif
 };
+#endif
 
 #if defined(CONFIG_FEC) || defined(CONFIG_FEC_MODULE)\
 	|| defined(CONFIG_FEC_L2SWITCH)
@@ -1285,3 +1288,12 @@
 						ARRAY_SIZE(mx28evk_eth_pins));
 #endif
 }
+
+#if defined(CONFIG_FB_MXS) || defined(CONFIG_FB_MXS_MODULE) || defined(CONFIG_FB_CHUMBYFBFW)
+void __init mx28evk_lcd_pins_init(void)
+{
+
+	mx28evk_init_pin_group(mx28evk_lcd_pins,
+						ARRAY_SIZE(mx28evk_lcd_pins));
+}
+#endif
diff -Nur linux-2.6.35.3.lastpatch/drivers/video/mxs/chumbyfbfw.c linux-2.6.35.3/drivers/video/mxs/chumbyfbfw.c
--- linux-2.6.35.3.lastpatch/drivers/video/mxs/chumbyfbfw.c	2011-06-22 16:55:56.611097250 +0800
+++ linux-2.6.35.3/drivers/video/mxs/chumbyfbfw.c	2011-07-15 16:00:03.999517689 +0800
@@ -48,7 +48,7 @@
 
 #include <linux/init.h>
 
-#define LOGO_OFFSET 0x44844000
+void __init mx28evk_lcd_pins_init(void);
 
 #define REGHW_LCDIF_CTRL1_RD() __raw_readl(lcdif_regbase + HW_LCDIF_CTRL1)
 #define REGHW_LCDIF_CTRL1_CLR(x) __raw_writel(x, lcdif_regbase + HW_LCDIF_CTRL1_CLR)
@@ -79,6 +79,8 @@
 #define REGHW_PXP_S0OFFSET_WR(x) __raw_writel(x, pxp_regbase + HW_PXP_S0OFFSET)
 #define REGHW_PXP_S0BACKGROUND_WR(x) __raw_writel(x, pxp_regbase + HW_PXP_S0BACKGROUND)
 
+unsigned char *logo_offset = NULL;
+
 // CHUMBY_fbsize
 //
 // Multiple framebuffers are presented as /dev/fbN.  This descriptor
@@ -1360,14 +1362,16 @@
     }
 
 
+    if (logo_offset != NULL)
+    {
 // CHUMBY_logo
     // Pre-copy the logo to both the screen (where the PXP will point to)
     // as well as pxp buffer 0.
     printk("Going to copy splash image from %p (%d bytes) to %p (not %p)\n",
-            (void *)LOGO_OFFSET, 320*240*2,
+            (void *)logo_offset, 320*240*2,
             data->virt_start, (void *)data->phys_start);
     {
-        char *old_fb = ioremap(LOGO_OFFSET, 320*240*2);
+        char *old_fb = ioremap((unsigned long) logo_offset, 320*240*2);
         if(old_fb) {
             memcpy(data->planes[0]->virt_start, old_fb, 320*240*2);
             memcpy(data->virt_start, old_fb, 320*240*2);
@@ -1377,6 +1381,7 @@
             CHLOG("Unable to call ioremap!");
     }
 // ! CHUMBY_logo
+    }
 
 
     chumbyfwfb_fix.smem_start = data->phys_start;
@@ -1470,6 +1475,7 @@
     }
 
 
+    mx28evk_lcd_pins_init();
 
     // Init the LCD.
     CHLOG("Calling init_lcdif()...\n");
@@ -1648,9 +1654,18 @@
     return 0;
 }
 
+static int __init logo_offset_setup(char *str)
+{
+  logo_offset = (unsigned char *) simple_strtoul(str, NULL, 16);
+  
+  printk("logo_offset = phy 0x%08x\n", (int) logo_offset);
+  
+  return 0;
+}
 
 MODULE_AUTHOR("Sean Cross <scross@chumby.com>");
 MODULE_DESCRIPTION("Chumby Wintergrasp Framebuffer Driver");
 MODULE_LICENSE("GPL");
 __setup("lcd_panel=", default_panel_select);
+__setup("logo_offset=", logo_offset_setup);
 
