From d0149d5a458912017f55ca194f6b2476dd9ac18b Mon Sep 17 00:00:00 2001
From: Sean Cross <sean@chumby.com>
Date: Wed, 4 May 2011 23:50:09 +0800
Subject: [PATCH 3/3] networkmanager: Make chumby plugin clean up after itself when an error happens

---
 system-settings/plugins/chumby/plugin.c |   22 ++++++++++++++++------
 1 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/system-settings/plugins/chumby/plugin.c b/system-settings/plugins/chumby/plugin.c
index c463a02..072cd09 100644
--- a/system-settings/plugins/chumby/plugin.c
+++ b/system-settings/plugins/chumby/plugin.c
@@ -132,23 +132,33 @@ reload_configs (gpointer data)
 	PLUGIN_PRINT("chumby", "Reloading configs...\n");
 
 
-        if(priv->iface_connections)
-		existing = priv->iface_connections;
-	else
-		existing = g_hash_table_new (g_str_hash, g_str_equal);
-	priv->iface_connections = g_hash_table_new (g_str_hash, g_str_equal);
-
 
 	/*parse the file and get the DOM */
 	doc = xmlReadFile(CONFIG_FILE, NULL, 0);
 
 	if (doc == NULL) {
 		PLUGIN_PRINT("chumby", "error: could not parse file %s\n", CONFIG_FILE);
+		xmlCleanupParser();
 		return;
 	}
 
 	/*Get the root element node */
 	root_element = xmlDocGetRootElement(doc);
+	if (!root_element) {
+		PLUGIN_PRINT("chumby", "error: could not find root element in " CONFIG_FILE "\n");
+		xmlFreeDoc(doc);
+		xmlCleanupParser();
+		return;
+	}
+
+
+        if(priv->iface_connections)
+		existing = priv->iface_connections;
+	else
+		existing = g_hash_table_new (g_str_hash, g_str_equal);
+	priv->iface_connections = g_hash_table_new (g_str_hash, g_str_equal);
+
+
 	for (cur_node = root_element->children; cur_node; cur_node = cur_node->next) {
 		if (cur_node->type == XML_ELEMENT_NODE && !strcmp((char *)cur_node->name, "configuration")) {
 			NMChumbyConnection *exported;
-- 
1.7.3.5

