diff -Nurd NetworkManager-0.8.1-stock/system-settings/plugins/chumby/plugin.c NetworkManager-0.8.1/system-settings/plugins/chumby/plugin.c
--- NetworkManager-0.8.1-stock/system-settings/plugins/chumby/plugin.c	2011-05-06 05:32:06.040142307 +0800
+++ NetworkManager-0.8.1/system-settings/plugins/chumby/plugin.c	2011-05-06 05:32:26.340142307 +0800
@@ -121,23 +121,33 @@
 	PLUGIN_PRINT("chumby", "Reloading configs...\n");
 
 
-        if(priv->iface_connections)
-		existing = priv->iface_connections;
-	else
-		existing = g_hash_table_new (g_str_hash, g_str_equal);
-	priv->iface_connections = g_hash_table_new (g_str_hash, g_str_equal);
-
 
 	/*parse the file and get the DOM */
 	doc = xmlReadFile(CONFIG_FILE, NULL, 0);
 
 	if (doc == NULL) {
 		PLUGIN_PRINT("chumby", "error: could not parse file %s\n", CONFIG_FILE);
+		xmlCleanupParser();
 		return;
 	}
 
 	/*Get the root element node */
 	root_element = xmlDocGetRootElement(doc);
+	if (!root_element) {
+		PLUGIN_PRINT("chumby", "error: could not find root element in " CONFIG_FILE "\n");
+		xmlFreeDoc(doc);
+		xmlCleanupParser();
+		return;
+	}
+
+
+        if(priv->iface_connections)
+		existing = priv->iface_connections;
+	else
+		existing = g_hash_table_new (g_str_hash, g_str_equal);
+	priv->iface_connections = g_hash_table_new (g_str_hash, g_str_equal);
+
+
 	for (cur_node = root_element->children; cur_node; cur_node = cur_node->next) {
 		if (cur_node->type == XML_ELEMENT_NODE && !strcmp((char *)cur_node->name, "configuration")) {
 			NMChumbyConnection *exported;
