diff -ur qt-everywhere-opensource-src-4.7.1-stock//src/3rdparty/harfbuzz/src/harfbuzz-global.h qt-everywhere-opensource-src-4.7.1/src/3rdparty/harfbuzz/src/harfbuzz-global.h
--- qt-everywhere-opensource-src-4.7.1-stock//src/3rdparty/harfbuzz/src/harfbuzz-global.h	2011-07-15 03:16:31.567721055 +0800
+++ qt-everywhere-opensource-src-4.7.1/src/3rdparty/harfbuzz/src/harfbuzz-global.h	2011-07-15 03:18:25.563718182 +0800
@@ -39,10 +39,6 @@
 #define HB_END_HEADER  /* nothing */
 #endif
 
-#if defined(__GNUC__) || defined(_MSC_VER)
-#define HB_USE_PACKED_STRUCTS
-#endif
-
 HB_BEGIN_HEADER
 
 #ifndef FALSE
