diff -Nurd lirc-0.8.7-stock/configure.ac lirc-0.8.7/configure.ac
--- lirc-0.8.7-stock/configure.ac	2011-06-03 12:46:23.604154998 +0800
+++ lirc-0.8.7/configure.ac	2011-07-02 22:40:39.326662153 +0800
@@ -174,6 +174,7 @@
 	(lirc_dev lirc_parallel) \
 	(lirc_dev lirc_sasem) \
 	(lirc_dev lirc_serial) \
+	(lirc_dev lirc_silvermoon) \
 	(lirc_dev lirc_sir) \
 	(lirc_dev lirc_streamzap) \
 	(lirc_dev lirc_wpc8769l)"
@@ -455,7 +456,7 @@
                           pcmak_usb, pctv, pixelview_bt878,
                           pixelview_pak, pixelview_pro, provideo,
                           realmagic, remotemaster, sa1100, samsung,
-                          sasem, sb0540, serial, silitek, sir, slinke,
+                          sasem, sb0540, serial, silvermoon, silitek, sir, slinke,
                           srm7500libusb, streamzap, tekram,
                           tekram_bt829, tira, tira_raw, ttusbir,
                           tuxbox, tvbox, udp, uirt2, uirt2_raw,
@@ -1334,6 +1335,10 @@
   lirc_driver="lirc_dev lirc_serial"
 fi
 
+if test "$driver" = "silvermoon"; then
+  lirc_driver="lirc_dev lirc_silvermoon"
+fi
+
 if test "$driver" = "silitek"; then
   lirc_driver="none"
   hw_module="hw_silitek.o serial.o"
@@ -1557,6 +1562,7 @@
 	lirc_parallel \
 	lirc_sasem \
 	lirc_serial \
+	lirc_silvermoon \
 	lirc_sir \
 	lirc_streamzap \
 	lirc_ttusbir \
@@ -1839,6 +1845,7 @@
   drivers/lirc_parallel/Makefile
   drivers/lirc_sasem/Makefile
   drivers/lirc_serial/Makefile
+  drivers/lirc_silvermoon/Makefile
   drivers/lirc_sir/Makefile
   drivers/lirc_streamzap/Makefile
   drivers/lirc_wpc8769l/Makefile
diff -Nurd lirc-0.8.7-stock/drivers/lirc_silvermoon/Makefile.am lirc-0.8.7/drivers/lirc_silvermoon/Makefile.am
--- lirc-0.8.7-stock/drivers/lirc_silvermoon/Makefile.am	1970-01-01 07:30:00.000000000 +0730
+++ lirc-0.8.7/drivers/lirc_silvermoon/Makefile.am	2011-07-02 22:35:31.842654785 +0800
@@ -0,0 +1,13 @@
+## $Id: Makefile.am,v 5.14 2004/04/25 16:29:27 lirc Exp $
+
+## Process this file with automake to produce Makefile.in
+
+## this is so that Automake includes the C compiling definitions, and
+## includes the source files in the distribution.
+EXTRA_PROGRAMS = automake_dummy
+automake_dummy_SOURCES = lirc_silvermoon.c
+
+## there is no *just* object file support in automake.  This is close enough
+module_DATA = lirc_silvermoon.o
+
+include ../Makefile.common
diff -Nurd lirc-0.8.7-stock/drivers/Makefile.am lirc-0.8.7/drivers/Makefile.am
--- lirc-0.8.7-stock/drivers/Makefile.am	2011-06-03 12:46:23.584155005 +0800
+++ lirc-0.8.7/drivers/Makefile.am	2011-07-02 22:34:18.722659774 +0800
@@ -22,6 +22,7 @@
 	lirc_parallel \
 	lirc_sasem \
 	lirc_serial \
+	lirc_silvermoon \
 	lirc_sir \
 	lirc_streamzap \
 	lirc_ttusbir \
