From 87247f724697dd74363f3b8e94ac35734d8786e9 Mon Sep 17 00:00:00 2001
From: Lei Wen <leiwen@marvell.com>
Date: Wed, 4 Mar 2009 19:13:27 +0800
Subject: [PATCH] pxa168: add switch download interfaces and set IP addr

Signed-off-by: Lei Wen <leiwen@marvell.com>
---
 board/pxa/common/tftp.c |   82 ++++++++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 81 insertions(+), 1 deletions(-)

diff --git a/board/pxa/common/tftp.c b/board/pxa/common/tftp.c
index b95cc5d..888d5ed 100644
--- a/board/pxa/common/tftp.c
+++ b/board/pxa/common/tftp.c
@@ -2,6 +2,86 @@
 #include <command.h>
 #include <net.h>
 
+int do_switch (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
+{
+	char *s;
+	char *net[2];
+	int now, i, original;
+	char *hostIP, *clientIP;
+
+#ifdef CONFIG_DRIVER_SMC91111
+	net[0] = "smc91111";
+#else
+	net[0] = NULL;
+#endif
+
+#ifdef CONFIG_USB_ETH
+	net[1] = "u2o-eth";
+#else
+	net[1] = NULL;
+#endif
+	/* get the pre-set host IP and client IP */
+	if ((s = getenv("ethact")) != NULL) {
+
+		for (i = 0; i < ARRAY_SIZE(net); i ++)
+			if (net[i] && strncmp(s, net[i], strlen(net[i])) == 0) {
+				now = i;
+				printf("Currently we use %s to download!!\n", net[i]);
+				break;
+			}
+
+		if (i == ARRAY_SIZE(net))
+			now = 0;
+	}
+	else
+		now = 0;
+
+	if (!net[now]) {
+		printf ("There is no interface defined!!!\n");
+		return -1;
+	}
+
+	original = now;
+	do {
+		/* set the NET to the next interface */
+		now += 1;
+		if (now >= ARRAY_SIZE(net))
+			now = 0;
+
+		if (net[now]) {
+			printf("***Set the interface to %s\n", net[now]);
+			setenv("ethact", net[now]);
+			break;
+		}
+	
+	} while (original != now);
+
+	if (original == now)
+		printf("The interface is still %s\n", net[original]);
+
+	/* change the host and client IP to specified addr*/
+	if (argc > 1) {
+		setenv ("serverip", argv[1]);
+		printf("reset host IP to %s\n", argv[1]);
+
+		if (argc > 2) {
+			setenv ("ipaddr", argv[2]);
+			printf("reset client IP to %s\n\n", argv[2]);
+		}
+		else
+			printf("client IP no change\n\n");
+	}
+	else
+		printf("IP no change\n\n");
+	return 0;
+}
+
+U_BOOT_CMD(
+	switch,	CONFIG_SYS_MAXARGS,	1,	do_switch,
+	"switch-\tswitch the download by USB or Ethernet, and config conresponsed IP\n",
+	"[hostIPaddr] [clientIP addr]\n"
+);
+
 int do_tftp (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
 {
 	char *s;
@@ -41,7 +121,7 @@ int do_tftp (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
 
 U_BOOT_CMD(
 	tftp,	CONFIG_SYS_MAXARGS,	1,	do_tftp,
-	"tftp-    download image via network using TFTP protocol\n",
+	"tftp-\tdownload image via network using TFTP protocol\n",
 	"[loadAddress] [[hostIPaddr:]filename]\n"
 );
 
-- 
1.6.0.4

