From 572142f22e9f1da7792841620047aafff4071cef Mon Sep 17 00:00:00 2001
From: Lei Wen <leiwen@marvell.com>
Date: Wed, 15 Jul 2009 16:33:36 +0800
Subject: [PATCH] nand: if erase fail, then relocate it on the fly

Signed-off-by: Lei Wen <leiwen@marvell.com>
---
 drivers/mtd/nand/nand_util.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/mtd/nand/nand_util.c b/drivers/mtd/nand/nand_util.c
index 7eb11ba..ffc4f35 100644
--- a/drivers/mtd/nand/nand_util.c
+++ b/drivers/mtd/nand/nand_util.c
@@ -144,6 +144,7 @@ int nand_erase_opts(nand_info_t *meminfo, const nand_erase_options_t *opts)
 		if (result != 0) {
 			printf("\n%s: MTD Erase failure: %d\n",
 			       mtd_device, result);
+			meminfo->block_markbad(meminfo, erase.addr);
 			continue;
 		}
 
-- 
1.6.0.4

