From a1f0e4d816f5db08d2c7738f0c816424ca1b7ef2 Mon Sep 17 00:00:00 2001
From: Mingliang Hu <mingliang.hu@marvell.com>
Date: Tue, 19 Jan 2010 18:44:03 +0800
Subject: [PATCH] pxa688: add op2 support for 624/400/200

Signed-off-by: Mingliang Hu <mingliang.hu@marvell.com>
---
 board/pxa/common/freq_ll.S |   23 +++++++++++++++++++++++
 1 files changed, 23 insertions(+), 0 deletions(-)

diff --git a/board/pxa/common/freq_ll.S b/board/pxa/common/freq_ll.S
index 0be414c..080bedc 100644
--- a/board/pxa/common/freq_ll.S
+++ b/board/pxa/common/freq_ll.S
@@ -145,6 +145,29 @@ freq_sram_start:
 	str	r7, [r6, #0x04]
 
 3:
+	cmp	r2, #0x2
+	bne	4f
+	@ select PLL2 frequency, 624MHz
+	ldr	r7, =0x086003e2
+	str	r7, [r5, #0x0414]
+	ldr	r7, =0x00FFFE00
+	str	r7, [r5, #0x0034]
+	ldr	r7, =0x00223A00
+	str	r7, [r5, #0x0034]
+	ldr	r7, =0x00223B00
+	str	r7, [r5, #0x0034]
+	ldr	r7, =0x286003e2
+	str	r7, [r5, #0x0414]
+	@ select clock source, PJ4-PLL2, SP-PLL1/2, AXI/DDR-PLL1
+	ldr	r7, =0x40800000
+	str	r7, [r5, #0x0008]
+	@ divider setting and frequency change request, core-624, ddr-400, axi-200
+	ldr	r7, =0x08fd8490
+	str	r7, [r6, #0x00]
+	ldr	r7, =0x78fd8490
+	str	r7, [r6, #0x04]
+
+4:
 	@
 	@ ddr re-calibration after frequency change
 	@
-- 
1.6.0.4

