From fbcb88ab0de4e8f38390698971eb2c3cb7f33848 Mon Sep 17 00:00:00 2001
From: Mingliang Hu <mingliang.hu@marvell.com>
Date: Wed, 10 Feb 2010 10:25:06 +0800
Subject: [PATCH] pxa688: update DDR setting for 400MHz

Signed-off-by: Mingliang Hu <mingliang.hu@marvell.com>
---
 board/pxa/common/freq_ll.S |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/board/pxa/common/freq_ll.S b/board/pxa/common/freq_ll.S
index 7a70cb4..5722e8e 100644
--- a/board/pxa/common/freq_ll.S
+++ b/board/pxa/common/freq_ll.S
@@ -300,7 +300,7 @@ freq_sram_start:
 	and	r7, r7, r8
 	str	r7, [r4, #0x200]
 
-	ldr	r7, =0x20004444
+	ldr	r7, =0x20004455
 	str	r7, [r4, #0x140]
 	ldr	r7, =0x13300559
 	str	r7, [r4, #0x1D0]
-- 
1.6.0.4

