diff -ur etc-stock/init.d/rcS etc/init.d/rcS
--- etc-stock/init.d/rcS	2010-10-23 05:14:27.388287261 +0000
+++ etc/init.d/rcS	2010-10-23 05:14:46.498285426 +0000
@@ -2,6 +2,11 @@
 # $Id: rcS 56474 2010-09-12 08:58:03Z scross $
 IS_RECOVERY=$(grep partition=recovery /proc/cmdline | wc -l)
 
+# In OE, we're not running under chumby's linuxrc.  Manually mount filesystems.
+mount /proc
+mount /sys
+mount -ttmpfs none /tmp
+
 # Determine what the brand is, and write that to /psp for the control panel
 for i in $(cat /proc/cmdline)
 do
