diff -ur src-stock/include/configs/aspenite.h src/include/configs/aspenite.h
--- src-stock/include/configs/aspenite.h	2011-05-18 16:46:52.000000000 +0000
+++ src/include/configs/aspenite.h	2011-05-18 16:52:31.000000000 +0000
@@ -244,16 +244,8 @@
     "echo \"Drawing something to the screen...\" ;"             \
     "snow init ${BP} ;"                                         \
     " ;"                                                        \
-    /* Determine if someone is touching the screen */           \
-    "echo \"Testing the touchscreen...\" ;"                     \
-    "TS=0 ; touchscreen || TS=1 ;"                              \
-    "if test ${TS} = 0 ;"                                       \
-    "    then snow load ${KF} /boot/logo_${brand}.gz ;"         \
-    "    else snow load ${KF} /boot/recovery_1_${brand}.gz ;"   \
-    "fi ;"                                                      \
-    "snow draw ;"                                               \
-    /*"snow start ;"*/                                          \
-    " ;"                                                        \
+    /* Ignore touchscreen for now */                            \
+    "TS=0 ;"                                                    \
     " ;"                                                        \
     /* Wait for the user to press Control-C */                  \
     "echo \"Press Control-C to enter a shell.\" ;"              \
@@ -262,10 +254,6 @@
     "    else echo \"Chumby shell\"; echo \"Type 'run bootcmd' to exit\"; exit ;" \
     "fi ;"                                                      \
     " ;"                                                        \
-    /* See if the user is still touching the screen */          \
-    "echo \"Testing the touchscreen again...\" ;"               \
-    "test ${TS} = 1 && TS=0 && touchscreen || TS=1 ;"           \
-    " ;"                                                        \
     /* If the user is pressing the screen, swap partitions */   \
     "if test ${TS} = 1 ;"                                       \
     "    then echo \"Booting to recovery shell...\" ;"          \
